File: mcp-code-intelligence-python/src/mcp_code_intel/graph/dependency_formatters.py
"""Dependency Formatters — tree/flat/graph output formats. KSA-179."""

from __future__ import annotations

import posixpath

from .models import DependencyNode, DependencyResult


def format_dependency_result(result: DependencyResult, fmt: str = "tree") -> dict:
    """Format a DependencyResult into the requested output format."""
    if fmt == "flat":
        return to_flat_format(result)
    if fmt == "graph":
        return to_graph_format(result)
    return to_tree_format(result)


def to_tree_format(result: DependencyResult) -> dict:
    """Build a tree representation of the dependency result."""
    by_depth: dict[int, list[DependencyNode]] = {}
    for node in result.results:
        by_depth.setdefault(node.depth, []).append(node)

    tree = {
        "file": result.root,
        "label": posixpath.basename(result.root),
        "depth": 0,
        "importedSymbols": [],
        "isExternal": False,
        "children": _build_children(by_depth, 1),
    }
    return {
        "root": result.root,
        "tree": tree,
        "cycles": result.cycles,
        "metadata": _metadata_dict(result),
    }


def to_flat_format(result: DependencyResult) -> dict:
    """Build a flat list representation."""
    return {
        "root": result.root,
        "direction": result.direction,
        "dependencies": [
            {
                "file": r.file,
                "depth": r.depth,
                "importedSymbols": r.imported_symbols,
                "isExternal": r.is_external,
            }
            for r in result.results
        ],
        "cycles": result.cycles,
        "metadata": _metadata_dict(result),
    }


def to_graph_format(result: DependencyResult) -> dict:
    """Build a nodes+edges graph representation."""
    nodes = [
        {"id": result.root, "label": posixpath.basename(result.root), "depth": 0, "isExternal": False},
        *[
            {"id": r.file, "label": posixpath.basename(r.file), "depth": r.depth, "isExternal": r.is_external}
            for r in result.results
        ],
    ]
    edges = [
        {"from": result.root, "to": r.file, "symbols": r.imported_symbols}
        for r in result.results
        if r.depth == 1
    ]
    return {"nodes": nodes, "edges": edges, "cycles": result.cycles, "metadata": _metadata_dict(result)}


def _build_children(by_depth: dict[int, list[DependencyNode]], depth: int) -> list[dict]:
    nodes = by_depth.get(depth, [])
    return [
        {
            "file": n.file,
            "label": posixpath.basename(n.file),
            "depth": n.depth,
            "importedSymbols": n.imported_symbols,
            "isExternal": n.is_external,
            "children": _build_children(by_depth, depth + 1),
        }
        for n in nodes
    ]


def _metadata_dict(result: DependencyResult) -> dict:
    m = result.metadata
    return {
        "totalNodes": m.total_nodes,
        "maxDepthReached": m.max_depth_reached,
        "truncated": m.truncated,
        "queryTimeMs": m.query_time_ms,
        "externalCount": m.external_count,
    }

File: backend/src/engine/graph/dependency-formatters.ts
/**
 * KSA-155: Dependency Formatters - tree/flat/graph output formats.
 */

import * as path from 'path';
import { DependencyNode, DependencyResult } from './dependency-graph-service.js';

export interface TreeNode {
  file: string;
  label: string;
  depth: number;
  importedSymbols: string[];
  isExternal: boolean;
  children: TreeNode[];
}

export function toTreeFormat(result: DependencyResult): object {
  const { root, results } = result;
  const byDepth = new Map<number, DependencyNode[]>();

  for (const node of results) {
    if (!byDepth.has(node.depth)) byDepth.set(node.depth, []);
    byDepth.get(node.depth)!.push(node);
  }

  const tree: TreeNode = {
    file: root,
    label: path.basename(root),
    depth: 0,
    importedSymbols: [],
    isExternal: false,
    children: buildChildren(byDepth, 1),
  };

  return { root, tree, cycles: result.cycles, metadata: result.metadata };
}

function buildChildren(byDepth: Map<number, DependencyNode[]>, depth: number): TreeNode[] {
  const nodes = byDepth.get(depth) || [];
  return nodes.map(n => ({
    file: n.file,
    label: path.basename(n.file),
    depth: n.depth,
    importedSymbols: n.importedSymbols,
    isExternal: n.isExternal,
    children: buildChildren(byDepth, depth + 1),
  }));
}

export function toFlatFormat(result: DependencyResult): object {
  return {
    root: result.root,
    direction: result.direction,
    dependencies: result.results.map(r => ({
      file: r.file,
      depth: r.depth,
      importedSymbols: r.importedSymbols,
      isExternal: r.isExternal,
    })),
    cycles: result.cycles,
    metadata: result.metadata,
  };
}

export function toGraphFormat(result: DependencyResult): object {
  const { root, results } = result;
  const nodes = [
    { id: root, label: path.basename(root), depth: 0, isExternal: false },
    ...results.map(r => ({
      id: r.file,
      label: path.basename(r.file),
      depth: r.depth,
      isExternal: r.isExternal,
    })),
  ];

  const edges: Array<{ from: string; to: string; symbols: string[] }> = [];
  // Build edges from results (each result is connected to root or parent at depth-1)
  for (const r of results) {
    if (r.depth === 1) {
      edges.push({ from: root, to: r.file, symbols: r.importedSymbols });
    }
  }

  return { nodes, edges, cycles: result.cycles, metadata: result.metadata };
}

export function formatDependencyResult(result: DependencyResult, format: string): object {
  switch (format) {
    case 'tree': return toTreeFormat(result);
    case 'flat': return toFlatFormat(result);
    case 'graph': return toGraphFormat(result);
    default: return toTreeFormat(result);
  }
}
